Shopify/ruby-lsp
# Add request cancellation

### Motivation

Closes #238. More context on the issue.

Process requests in a `Thread` worker to allow the main thread to manage a queue of requests.

### Implementation

The first commit adds the concurrency and queue mechanism. The second commit refactors the code to separate running requests from writing the result to the IO. We can only cancel while processing requests, but writing to the IO must always happen.

This PR
1. Allows us to configure which requests can run in parallel and which ones must run in the main thread
2. Adds a `Queue` to process requests in a worker
3. Adds the capability to cancel requests

### Automated Tests

Added tests for the queue. After the refactors, it is not longer possible to test the example that existed in `handler_test`, so I removed that file.

### Manual Tests

There is not a lot of specific things to check. Basically just a smoke test trying things out and making sure nothing is breaking.
1. Start the LSP on this branch
2. Play around with editing files, using folding ranges, document symbols, etc
3. Verify that requests continue to work as expected
